invenio_rdm_records/services/services.py

```diff
@@ -21,6 +29,299 @@ class RDMRecordService(RecordDraftService):
+        resolved.
+        """
+        try:
+            record = self.record_cls.pid.resolve(id_, registered_only=False)
+            parent = record.parent
+        except Exception:  # TODO NoResultFoundException?
+            record = None
+            parent = self.record_cls.parent_record_cls.pid.resolve(
+                id_, registered_only=False
+            )
+
+        return record, parent
+
+    def _validate_secret_link_expires_at(
+        self, expires_at, is_specified=True, secret_link=None
+    ):
+        """Validate the given expiration date.
+
+        If a ``secret_link`` is specified, the validity of setting its
+        expiration date to ``expires_at`` will be checked additionally.
+        The ``is_specified`` flag hints at if the value of ``expires_at``
+        was set in the given data, or if it was omitted (which makes a
+        difference in patch operations).
+        """
+        if expires_at and is_specified:
+            # if the expiration date was specified, check if it's in the future
+            expires_at = arrow.get(expires_at).to("utc").datetime
+            expires_at = expires_at.replace(tzinfo=None)
+
+            if expires_at < datetime.utcnow():
+                raise ValidationError(
+                    message=_("Expiration date must be set to the future"),
+                    field_name="expires_at",
+                )
+
+        if secret_link is not None:
+            # if we're updating an existing secret link, we need to do some
+            # more checks...
+
+            # we interpret explicitly setting 'expires_at = null/None' as
+            # removing the expiration date (semantically different from not
+            # specifying an 'expires_at' value at all, at least for updates)
+            increases_expiration = (
+                is_specified and not expires_at and secret_link.expires_at
+            ) or (
+                expires_at
+                and secret_link.expires_at
+                and expires_at > secret_link.expires_at
+            )
+
+            if increases_expiration:
+                # it's not a problem to reduce the validity of a token (*),
+                # but increasing its lifespan would require a new signature,
+                # and thus a new token
+                # (*) in that case, the permission generator will still say
+                #     no, even if the signature is still valid
+                raise ValidationError(
+                    message=_("Cannot postpone expiration of links"),
+                    field_name="expires_at",
+                )
+
+            elif expires_at and expires_at < datetime.utcnow():
+                raise ValidationError(
+                    message=_("Expiration date must be set to the future"),
+                    field_name="expires_at",
+                )
+
+        return expires_at
+
+    def _index_related_records(self, record, parent):
```

Comment:
Probably this should live in drafts-resources as all the logic related to parents is there.